### PR TITLE
SparseCsrCUDA: cuDSS backend for linalg.solve

### LINKED ISSUE #69538: cuSOLVER backend for Sparse CSR direct linear solvers
## 🚀 Feature

cuSolverSP (part of cuSOLVER) provides linear solver, least-square solver routines for matrices in CSR format. We can use it as a backend for `torch.linalg.solve` and `torch.linalg.lstsq`.

We are working on improving coverage of Sparse CSR matrices in PyTorch. Solving systems of linear equations is one of the fundamental operations in computational linear algebra.

From cuSOLVER docs:
> The high-level API is designed for ease-of-use, so it allocates any required memory under the hood automatically. If the host or GPU system memory is not enough, an error is returned.

Unfortunately, it means these functions do not allow passing externally allocated buffers, and paths with inputs on GPU are not implemented (there remarks for relevant functions "only CPU (Host) path is provided").

- [ ] `torch.linalg.solve`
    Use [`cusolverSp<t>csrlsvlu`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolver-lt-t-gt-csrlsvlu). 

- [ ] `torch.linalg.lstsq`
    Use [`cusolverSp<t>csrlsqvqr`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolver-lt-t-gt-csrlsqvqr)

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel @jianyuh @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by zitongzhan**
This PR switches to cuDSS library and has the same purpose of #127692, which is to add Sparse CSR tensor support to linalg.solve.
Fixes #69538

Minimum example of usage:
```
import torch

if __name__ == '__main__':
    spd = torch.rand(4, 3)
    A = spd.T @ spd
    b = torch.rand(3).to(torch.float64).cuda()
    A = A.to_sparse_csr().to(torch.float64).cuda()

    x = torch.linalg.solve(A, b)
    print((A @ x - b).norm())

```
cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip @nWEIdia @lezcano @kvoronin @IvanYashchuk @WindowsXp-Beta

### CODE REVIEW

File: aten/src/ATen/native/cuda/linalg/CudssHandlePool.cpp
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/linalg/CudssHandlePool.cpp:1]: Note to reviewers:
This is simply a copy following other pool handle examples:
* https://github.com/pytorch/pytorch/blob/005ae01d6ae4a57af21bc3b8edf2ca21e5fed73d/aten/src/ATen/cuda/CublasHandlePool.cpp
* https://github.com/pytorch/pytorch/blob/005ae01d6ae4a57af21bc3b8edf2ca21e5fed73d/aten/src/ATen/cuda/CuSparseHandlePool.cpp
* https://github.com/pytorch/pytorch/blob/005ae01d6ae4a57af21bc3b8edf2ca21e5fed73d/aten/src/ATen/native/cuda/linalg/CusolverDnHandlePool.cpp

File: aten/src/ATen/native/native_functions.yaml
Comment (IvanYashchuk) [inline aten/src/ATen/native/native_functions.yaml:14179]: This change would break the dispatcher composability (for example I think with these dispatch keys autograd is broken because there's no rule for `linalg_solve`, only for `_linalg_solve_ex`:
https://github.com/pytorch/pytorch/blob/313eec02cc6bbdf3310801cfc04d750ba5539eef/tools/autograd/derivatives.yaml#L1562-L1565
`linalg_solve` is defined here:
https://github.com/pytorch/pytorch/blob/005ae01d6ae4a57af21bc3b8edf2ca21e5fed73d/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1934-L1939

I think it should be marked as `CompositeImplicitAutograd`, a little bit of description of what it does can be found here https://github.com/pytorch/pytorch/blob/005ae01d6ae4a57af21bc3b8edf2ca21e5fed73d/aten/src/ATen/native/README.md#dispatch
```suggestion
    CompositeImplicitAutograd: linalg_solve
```

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (IvanYashchuk) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:721]: Have you hit a problem with a matrix for `b`? cuDSS documentation says the right-hand-side argument can be a matrix.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (IvanYashchuk) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:757]: Column- or row-major layout should be checked from the strides of `b` and `x`. The leading dimension argument shouldn't be `b.size(0)` but `max(1, b.stride(0 or 1 based on layout)` to allow a wider variety of strided inputs.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (IvanYashchuk) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:820]: Strides of `b` should be checked if they are compatible with `_apply_sparse_csr_linear_solve` if not a copy into a column-major contiguous tensor needs to be done.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (IvanYashchuk) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:757]: Oh, cuDSS documentation says it only supports column-major layout and the leading dimensions equal to exactly the number of rows.
https://docs.nvidia.com/cuda/cudss/functions.html#cudssmatrixcreatedn

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (zitongzhan) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:721]: I haven't encountered before, but I'll stay tuned on any future update on cuDSS.

File: aten/src/ATen/native/cuda/linalg/CudssHandlePool.cpp
Comment (zitongzhan) [reply aten/src/ATen/native/cuda/linalg/CudssHandlePool.cpp:1]: Right, it's copied from [/aten/src/ATen/native/cuda/linalg/CusolverDnHandlePool.cpp](https://github.com/pytorch/pytorch/blob/005ae01d6ae4a57af21bc3b8edf2ca21e5fed73d/aten/src/ATen/native/cuda/linalg/CusolverDnHandlePool.cpp) with minor changes. We could consider simplifying these handlers in the future.

File: aten/src/ATen/native/native_functions.yaml
Comment (zitongzhan) [reply aten/src/ATen/native/native_functions.yaml:14179]: I tried setting `CompositeImplicitAutograd` dispatch key for `linalg_solve` function. However, it caused the `SparseCsrCUDA: _sparse_csr_linear_solve` dispatch rule no longer effective. In other words, all `torch.linalg.solve` function calls with CUDA Sparse CSR tensor inputs are also being routed to `at:: linalg_solve`. Any workaround on this?

**Comment by AlexanderAKalinkin** [comment]
Hi All,
new version of cuDSS was [released](https://developer.nvidia.com/cudss-downloads). I recommend to try the new version because of performance improvement and some API changes. 
Thanks,
Alex

**Comment by zitongzhan** [comment]
> Hi All, new version of cuDSS was [released](https://developer.nvidia.com/cudss-downloads). I recommend to try the new version because of performance improvement and some API changes. Thanks, Alex

Thanks! I reviewed the release note, and fortunately there's no breaking changes impacting this PR.

File: aten/src/ATen/native/native_functions.yaml
Comment (huydhn) [reply aten/src/ATen/native/native_functions.yaml:14179]: @zitongzhan Hi Zitong, I see that you have book an OH with us (PT Dev Infra) this Friday (July 5th).  However, please keep in mind that we could only help unblock CI issues on your PR.  A technical question like the one you have here should better be answered by experts like @IvanYashchuk and @eqy

**Comment by zitongzhan** [comment]
Now the function is being hooked to both `torch.sparse.solve` and `torch.linalg.solve`

File: aten/src/ATen/native/native_functions.yaml
Comment (zitongzhan) [reply aten/src/ATen/native/native_functions.yaml]: Resolved via [6853fa6](https://github.com/pytorch/pytorch/pull/129856/commits/6853fa60a830bdf4019a1b3d7a1f4547163c5879)

**Comment by zitongzhan** [comment]
I added the cudss package to the Ubuntu CUDA docker, so that cudss function calls in this PR could be tested.

File: test/allowlist_for_publicAPI.json
Comment (albanD) [inline test/allowlist_for_publicAPI.json:899]: Adding new entry to this file is not ok. Please make sure your API is properly exposed.

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/allowlist_for_publicAPI.json
Comment (zitongzhan) [reply test/allowlist_for_publicAPI.json]: I see what's going on. Problem fixed in the commit below.

**Comment by zitongzhan** [comment]
@IvanYashchuk Would you like to review this PR again? I checked your comments and fixed CI issues. Sorry about the delay.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (zitongzhan) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu]: Updated in [fed22ad](https://github.com/pytorch/pytorch/pull/129856/commits/fed22ad2a3b533948b5f6ed779f0fa0a58eeb0de)

File: test/test_sparse_csr.py
Comment (lezcano) [inline test/test_sparse_csr.py:3494]: Just instantiate this test on the CUDA device and skip it as well at the top if you are not testing on CUDA.

**Review by lezcano — COMMENTED** [review]
@amjames could you review this one?

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (amjames) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:724]: Checks are needed to enforce the size/stride requirements on `x`. There should also be a guard against `x` or `b` having non-strided layout.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (amjames) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:757]: It looks like all of the struct creation and execute calls will have status returns that should be checked.

**Review by amjames — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/native_functions.yaml
Comment (zitongzhan) [reply aten/src/ATen/native/native_functions.yaml:14217]: I am writing it in this way because `linalg_solve` is already (implicitly) using the `CompositeImplicitAutograd` dispatch key. `CompositeImplicitAutograd` has higher priority than `SparseCsrCUDA`, so adding `SparseCsrCUDA ` dispatch rule under `linalg_solve` would have no effect in my previous attempts.

File: aten/src/ATen/native/cuda/linalg/CUDASolver.cpp
Comment (zitongzhan) [reply aten/src/ATen/native/cuda/linalg/CUDASolver.cpp]: fixed by https://github.com/pytorch/pytorch/pull/129856/commits/7e40ed3e4d0a295d17ea6324ed121bc5284de6d4

File: test/test_sparse_csr.py
Comment (zitongzhan) [reply test/test_sparse_csr.py]: fixed by https://github.com/pytorch/pytorch/pull/129856/commits/1dba412a66fab54bdfbfab51555d37705a4c73bb

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (zitongzhan) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu]: fixed by https://github.com/pytorch/pytorch/pull/129856/commits/51614e9ccc67bfb214ea6f1f4605997fd8843d53

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (zitongzhan) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:731]: fixed by https://github.com/pytorch/pytorch/pull/129856/commits/9c61f3dd0bd1d09bfafe198f0b6c0ae26a029050

**Comment by zitongzhan** [comment]
@amjames Thanks for the comments. I made the above changes.

File: test/test_sparse_csr.py
Comment (zitongzhan) [reply test/test_sparse_csr.py]: Thanks! Updated in https://github.com/pytorch/pytorch/pull/129856/commits/4238b479911f3f542b03cf785f97a26685243821

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (amjames) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:811]: Nit: We don't need a conditional here. `Tensor b_copy = b.contiguous()` will be a no op if the tensor is already contiguous.

**Review by amjames — APPROVED** [review]
Looks good from my end.

@lezcano or someone else with permissions will need to approve before merge though.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (zitongzhan) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu]: Updated in https://github.com/pytorch/pytorch/pull/129856/commits/882d5a6637edb6fd419a0f5c0ae80f064d6df12d

**Review by lezcano — APPROVED** [review]
didn't review. @amjames approved

**Review by lezcano — COMMENTED** [review]
also, does backward just throw an error when doing `linalg.solve(csr, dense)` or what is its behaviour?

And can you add a short line to `line.solve` docs mentioning that it works with CSR tensors?

**Comment by zitongzhan** [comment]
> also, does backward just throw an error when doing `linalg.solve(csr, dense)` or what is its behaviour?
> 
> And can you add a short line to `line.solve` docs mentioning that it works with CSR tensors?

It will throw the error in the backward pass `RuntimeError: derivative for aten::_spsolve is not implemented`
Sure, added https://github.com/pytorch/pytorch/pull/129856/commits/60f7649a3f9c2ad22f1064993c993732ceca28e6

File: torch/linalg/__init__.py
Comment (lezcano) [inline torch/linalg/__init__.py:2135]: This comment is stale. We should remove it in a different PR. Also, the note that this function works with CSR tensors might be better placed in the paragraph where we discuss dtype and batched support above. Not for this PR tho.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (kvoronin) [inline aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:763]: Hi! Could you comment on the necessity of this? Are there particular test cases which cannot pass with the the default algorithm (same as setting `CUDSS_ALG_DEFAULT` or not setting anything at all)
The default reordering algorithm (METIS-based) has better perf in most of the cases, sometimes significantly (but admittedly can have much more perturbed pivots and less accuracy sometimes).

Also, if trying the default algorithm, I suggest making 2 steps of iterative refinement by default (which can help with perturbed pivots for `CUDSS_ALG_DEFAULT` reordering algorithm) via `cudssConfigSet()` with `CUDSS_CONFIG_IR_N_STEPS`.
For `CUDSS_ALG_1` reordering, iterative refinement likely is not needed.

File: cmake/Modules/FindCUDSS.cmake
Comment (kvoronin) [inline cmake/Modules/FindCUDSS.cmake:26]: With cudss 0.3.0, we have renamed the main header from `cuDSS.h` into `cudss.h` (unpleasant change but encessary to align with other libraries). Should we correct it here as well?

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (zitongzhan) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:763]: When I was using the cudss 0.2 version, the `cudssExecute(handle, CUDSS_PHASE_ANALYSIS, config, cudss_data, A_mt, x_mt, b_mt)` statement using `CUDSS_ALG_DEFAULT` always return a status of internal error and fail the `TORCH_CUDSS_CHECK`, even though the solution it gives is correct. 
Looks like the bug has been fixed in 0.3 so I am going to switch back to `CUDSS_ALG_DEFAULT`.
Thanks for the input!

File: cmake/Modules/FindCUDSS.cmake
Comment (zitongzhan) [reply cmake/Modules/FindCUDSS.cmake]: Yes, corrected in https://github.com/pytorch/pytorch/pull/129856/commits/9d3d6ffccc717ddac473bc467aacb6c474d80c39

**Comment by huydhn** [comment]
@zitongzhan We have a sanity check in Dr.CI where it only accepts up to 10 unrelated failures.  If there are more than that (19 in this case), you will need to force merge it explicitly.  As the PR has already been approved, feel free to force merge it.  As there are multiple docker failures, they look like infra outage

**Comment by lezcano** [comment]
@huydhn this PR touches some docker files and it has some failing docker tests. Could you review that bit, and if that bit looks alright to you, do you mind force-merging?

**Review by huydhn — APPROVED** [review]
The docker failures are unrelated and I think they are ok now after recent rebase/retry

**Comment by vadimkantorov** [comment]
Does this implement analogue of scipy.sparse.linalg.spsolve? 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.spsolve.html

**Comment by kvoronin** [comment]
FWIW, cuDSS also supports non-uniform and uniform (same sparsity pattern) batches for the sparse (system) matrices. in The language of sparse direct solvers, batch RHS is just called multiple rhs support (and is enabled usually in all solvers at the very beginning, only performance optimizations may lag).

**Pull request closed without merging**
